analyzers/tests/SonarAnalyzer.UnitTest/Rules/Utilities/TokenTypeAnalyzerTest.ClassifierTestHarness.cs

```diff
@@ -0,0 +1,109 @@
+        {
+            var (tree, model, expectedTokens) = ParseTokens(code, ignoreIssues);
+            var root = tree.GetRoot();
+            model = allowSemanticModel ? model : new Mock<SemanticModel>(MockBehavior.Strict).Object; // The Mock will throw if the semantic model was used.
+            var tokenClassifier = new SonarAnalyzer.Rules.CSharp.TokenTypeAnalyzer.TokenClassifier(model, false);
+            var triviaClassifier = new SonarAnalyzer.Rules.CSharp.TokenTypeAnalyzer.TriviaClassifier();
+            expectedTokens.Should().SatisfyRespectively(expectedTokens.Select((Func<ExpectedToken, Action<ExpectedToken>>)(e => expected =>
+            {
+                var expectedLineSpan = tree.GetLocation(expected.Postion).GetLineSpan();
+                var because = $$"""token with text "{{expected.TokenText}}" at position {{expectedLineSpan}} was marked as {{expected.TokenType}}""";
+                var (actualLocation, classification) = FindActual();
+                if (classification == null)
+                {
+                    because = $$"""classification for token with text "{{expected.TokenText}}" at position {{expectedLineSpan}} is null""";
+                    expected.TokenType.Should().Be(TokenType.UnknownTokentype, because);
+                    actualLocation.SourceSpan.Should().Be(expected.Postion, because);
+                }
+                else
+                {
+                    classification.Should().Be(new TokenTypeInfo.Types.TokenInfo
+                    {
+                        TokenType = expected.TokenType,
+                        TextRange = new TextRange
+                        {
+                            StartLine = expectedLineSpan.StartLinePosition.Line + 1,
+                            StartOffset = expectedLineSpan.StartLinePosition.Character,
+                            EndLine = expectedLineSpan.EndLinePosition.Line + 1,
+                            EndOffset = expectedLineSpan.EndLinePosition.Character,
+                        },
+                    }, because);
+                }
+
+                (Location, TokenTypeInfo.Types.TokenInfo TokenInfo) FindActual()
+                {
+                    if (expected.TokenType == TokenType.Comment)
+                    {
+                        var trivia = root.FindTrivia(expected.Postion.Start);
+                        return (tree.GetLocation(trivia.FullSpan), triviaClassifier.ClassifyTrivia(trivia));
+                    }
+                    else
+                    {
+                        var token = root.FindToken(expected.Postion.Start);
+                        var f = () => tokenClassifier.ClassifyToken(token);
+                        var tokenInfo = f.Should().NotThrow($"semanticModel should not be queried for {because}").Which;
+                        return (token.GetLocation(), tokenInfo);
+                    }
```

Comment:
I would extract the action given to the `SatisfyRespectively` in another method or local function - otherwise, I find it very confusing.